Fix wrong default download directory in Windows. Closes #2625.

Chocobo1 opened the pull request:
This should be the correct default path for windows vista (and above), but not for windows xp (which MS doesn't support anymore...)

ngosang commented:
I prefer to maintain compatibility with XP. There must be a solution to check the version of Windows.

Chocobo1 commented:
Can testers/devs help me verify on winXP & win7?
I will squash commits into one after it's verified.

ngosang commented:
I'm a Linux user. I think @DoumanAsh can help us.

DoumanAsh commented:
I have win8.1 but i didn't make nay build tools for qBT in win.
I have only virtual box with Unix.

Starting from Win Vista we do have default Downloads folder in:
C:\Users\[user]\Downloads

So if HomeLocation returns directory of the current user then i think it is correct. But we need someone to build as it will take me forever to set-up build environment in windows :)

Chocobo1 commented:
Commits squashed.

Chocobo1 commented:
Rebased, #3277 is also ready.

Chocobo1 commented:
@sledgehammer999 , care to take a look at this?

ngosang commented:
I didn't test but looks good.

davispuh commented:
That's such a hack... For all OS'es you should use `QStandardPaths::standardLocations(QStandardPaths::DownloadLocation)`  (that's the point of Qt that it abstracts all mess away).
On Windows since Qt 5.3 this will return correct Download path.

DoumanAsh commented:
@davispuh 
It raises issue of using QT5.3 cuz as of current QT4 is used for official builds

Chocobo1 commented:
Rebased. I did a proper recheck on this PR and simplified some code.

I hardcoded "Downloads" instead of using the translated string, because it works more correct when pressing this button: ![1](https://cloud.githubusercontent.com/assets/9395168/9054360/edb43f40-3ab2-11e5-8de0-7145fcc91d4c.png)
Before it goes into "User" folder. Now it goes into the correct "Downloads" folder.

> On Windows since Qt 5.3 this will return correct Download path.

Seems you're right.

davispuh commented:
I recommend cleaning it up more and fix OS X.
So for backwards compatibility could do like this

```c++
QString Utils::Fs::QDesktopServicesDownloadLocation()
{
#if (QT_VERSION >= QT_VERSION_CHECK(5, 0, 0))
    return QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);
#else // for Qt 4
    // remove this section once Qt 4 isn't supported anymore.
#if !defined(Q_OS_LINUX) // everything non-Linux
    auto location = QDesktopServices::HomeLocation;
#if defined(Q_OS_WIN) // special case for WinXP, remove once not supported
    if (QSysInfo::windowsVersion() <= QSysInfo::WV_XP) { // Windows XP
        location = QDesktopServices::DocumentsLocation;
    };
#endif // defined(Q_OS_WIN)
   return QDir(QDesktopServices::storageLocation(location)).absoluteFilePath("Downloads");
#else // for Linux
    QString save_path;
    // Default save path on Linux
    QString config_path = QString::fromLocal8Bit(qgetenv("XDG_CONFIG_HOME").constData());
    if (config_path.isEmpty())
        config_path = QDir::home().absoluteFilePath(".config");

    QString user_dirs_file = config_path + "/user-dirs.dirs";
    if (QFile::exists(user_dirs_file)) {
        QSettings settings(user_dirs_file, QSettings::IniFormat);
        // We need to force UTF-8 encoding here since this is not
        // the default for Ini files.
        settings.setIniCodec("UTF-8");
        QString xdg_download_dir = settings.value("XDG_DOWNLOAD_DIR").toString();
        if (!xdg_download_dir.isEmpty()) {
            // Resolve $HOME environment variables
            xdg_download_dir.replace("$HOME", QDir::homePath());
            save_path = xdg_download_dir;
            qDebug() << Q_FUNC_INFO << "SUCCESS: Using XDG path for downloads: " << save_path;
        }
    }

    // Fallback
    if (!save_path.isEmpty() && !QFile::exists(save_path)) {
        QDir().mkpath(save_path);
    }

    if (save_path.isEmpty() || !QFile::exists(save_path)) {
        save_path = QDir::home().absoluteFilePath(QCoreApplication::translate("fsutils", "Downloads"));
        qDebug() << Q_FUNC_INFO << "using" << save_path << "as fallback since the XDG detection did not work";
    }

    return save_path;
#endif // !defined(Q_OS_LINUX) 
#endif // QT_VERSION
}
```

sledgehammer999 commented on src/core/utils/fs.cpp:
```diff
@@ -453,6 +453,18 @@ QString Utils::Fs::QDesktopServicesDownloadLocation()
+    else
+        return QDir(QDesktopServices::storageLocation(QDesktopServices::HomeLocation)).absoluteFilePath("Downloads");
+#else
+    return QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);
```
This potentially could be used for Linux too. I don't know if Qt5 respects `XDG_DOWNLOAD_DIR` though. This isn't high prio, since the other code "just works".
However, this presents an issue with qt5 and windows XP. It points to `C:\Documents and Settings\<username>\My Documents` instead of `C:\Documents and Settings\<username>\My Documents\Downloads` like we do now.
Why? Well, Windows XP doesn't actually have a downloads folder. It is a de fact standatd to put them to "Downloads" inside "My Documents".

Btw, do you have a non-english Windows Vista+ to test? I think the "Downloads" gets translated in other languages. So does this function still point to the correct folder?

davispuh replied:
Qt5 does use XDG_DOWNLOAD_DIR for QStandardPaths::DownloadLocation
Also in Windows since Vista (no idea about XP) folder's aren't translated on filesystem, but it's done by Explorer. So on disk you've "Downloads" not translated, but if you look with Explorer you'll see it translated while actually on disk it's not.

sledgehammer999 replied:
Nice info.
Then @Chocobo1 can you use `QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);` for all OS's if program is compiled with Qt5?

Chocobo1 replied:
> So does this function still point to the correct folder?

On Vista+, the function is working correctly as davispuh said.
On XP, the `Downloads` folder is using translated string unlike on Vista+. Unfortunately I couldn't compile a binary to run in XP and test the function.

> Then @Chocobo1 can you use...

Will change.

Chocobo1 commented:
Updated.

BTW, in `fs.cpp`, shouldn't
```c++
#include "misc.h"
#include "fs.h"
```
be
```c++
#include "fs.h"
#include "misc.h"
```
?
Also in `misc.cpp`, `misc.h` should be included before the `core/*` ones, no?

sledgehammer999 commented on src/core/utils/fs.cpp:
```diff
@@ -439,18 +439,25 @@ QString Utils::Fs::QDesktopServicesCacheLocation()
 #endif
+
+#if defined(Q_OS_WIN)
+    if (QSysInfo::windowsVersion() <= QSysInfo::WV_XP)  // Windows XP
```
This needs to be done for qt5+win XP too.
I guess this is the final change. So please update your v3_2_x PR too.

Chocobo1 replied:
Done.

sledgehammer999 commented on src/core/utils/fs.cpp:
```diff
@@ -439,18 +439,30 @@ QString Utils::Fs::QDesktopServicesCacheLocation()
     // as long as it stays WinXP like we do the same on OS/2
     // TODO: Use IKnownFolderManager to get path of FOLDERID_Downloads
     // instead of hardcoding "Downloads"
     // Unfortunately, this would break compatibility with WinXP
```
Delete this comment. It doesn't belong in the Q_OS_OS2 case.

sledgehammer999 replied:
No don't delete. Transfer it below to WINDOWS case. Or make this WINDOWS case and below OS/2 case.

Chocobo1 replied:
Move this whole comment block to line 461?

sledgehammer999 replied:
Yes.

Chocobo1 replied:
Done.

sledgehammer999 commented:
Perfect. Thank you.
